### PR TITLE
[WIP] Run kube-scheduler as non-root in kubeadm.

### PR DESCRIPTION
**Pull request by vinayakankugoyal**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -111,9 +112,10 @@ require (
 	gopkg.in/gcfg.v1 v1.2.0
 	gopkg.in/square/go-jose.v2 v2.2.2
 	gopkg.in/yaml.v2 v2.2.8
+	gopkg.in/yaml.v3 v3.0.0-20200615113413-eeeca48fe776 // indirect
 	k8s.io/api v0.0.0
 	k8s.io/apiextensions-apiserver v0.0.0
-	k8s.io/apimachinery v0.0.0
+	k8s.io/apimachinery v0.19.2
```
Comment (BenTheElder) [inline go.mod:118]: this looks wrong, it should be the version symlinked from staging/, we override this with a `replace`.

use `hack/update-vendor.sh` when modifying deps.

File: cmd/kubeadm/app/apis/kubeadm/v1beta1/defaults.go [hunk 2]
```diff
@@ -101,6 +101,12 @@ func SetDefaults_ClusterConfiguration(obj *ClusterConfiguration) {
 	SetDefaults_DNS(obj)
 	SetDefaults_Etcd(obj)
 	SetDefaults_APIServer(&obj.APIServer)
+	SetDefaults_Scheduler(&obj.Scheduler)
+}
+
+func SetDefaults_Scheduler(obj *ControlPlaneComponent) {
+	obj.UID = 2000
+	obj.GID = 2000
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/v1beta1/defaults.go:109]: existing kubeadm APIs are locked - v1beta1, v2beta2 and new fields will not be added to them.
all new fields must be added to v1beta3.

File: cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go [hunk 3]
```diff
@@ -126,6 +126,10 @@ type ControlPlaneComponent struct {
 
 	// ExtraVolumes is an extra set of host volumes, mounted to the control plane component.
 	ExtraVolumes []HostPathMount `json:"extraVolumes,omitempty"`
+
+	UID int64
+
+	GID int64
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go:132]: i know this is a draft PR, but i'm not a fan of letting the user control this.
if we are going to migrate one component at at time to non-root that seems OK, but ideally we should just have all CP components migrate to rootless without API changes:
https://github.com/kubernetes/kubeadm/issues/1367

File: go.mod (hunk 1 above)
Comment (neolit123) [reply go.mod:118]: the go.mod / vendor changes in this PR are not needed.

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go [hunk 4]
```diff
@@ -77,29 +77,53 @@ func CreateJoinControlPlaneKubeConfigFiles(outDir string, cfg *kubeadmapi.InitCo
...
+	if err := os.Chown(file, int(cfg.ControllerManager.UID), int(cfg.ControllerManager.GID)); err != nil {
+		return err
+	}
+
+	if err := createKubeConfigFiles(outDir, cfg, kubeadmconstants.SchedulerKubeConfigFileName); err != nil {
+		return err
 	}
+	fmt.Printf("[kubeconfig] setting owner of kubeconfig file %s to uid: %d gid: %d\n", kubeadmconstants.AdminKubeConfigFileName, cfg.Scheduler.UID, cfg.Scheduler.GID)
+	file = filepath.Join(outDir, kubeadmconstants.SchedulerKubeConfigFileName)
+	if err := os.Chown(file, int(cfg.Scheduler.UID), int(cfg.Scheduler.GID)); err != nil {
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:105]: chown-ing a kubeconfig file is not in scope of https://github.com/kubernetes/kubeadm/issues/1367

Comment (BenTheElder) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:105]: This is only the file we just created?
What do you propose instead?

Comment (neolit123) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:105]: around 2018 there was a discussion whether kubeadm should chown the files it creates, and everybody voted no.
is there are related issue to this PR?

Comment (BenTheElder) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:105]: there's no mention of chown in the linked issue, can you provide some more context for the 2018 discussion?
I would expect the components to own their own config.

Comment (neolit123) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:105]: > there's no mention of chown in the linked issue, can you provide some more context for the 2018 discussion?

i had a PR at some point to make the kubeadm admin.conf be chowed for a target UID / GID which helps the last manual step after `kubeadm init` finishes and kubeconfigs must be copied to `~/.kube`.
the idea was dropped with the argument that users can do whatever they want with kubeconfigs and kubeadm should not be in charge of chowing or moving them.

notably chown doesn't work for Windows, so the current code in this PR is not portable.

> I would expect the components to own their own config.

this writes a kubeconfig on disk and chowns it for a certain UID / GID, which is strange and i'm confused why its needed. on a kubeadm host /etc/kubernetes/* is owned by root. https://github.com/kubernetes/kubeadm/issues/1367 is about moving components running in containers away from root.

Comment (vinayakankugoyal) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:105]: If a component is running as non root (like kubescheduler in this case) and mounts hostVols (like kubeconfig file in this case) it needs permissions to access those files. So when we run kube-scheduler as uid 2000 the scheduler.conf file must be accessible to user 2000, hence the chown is required.

Comment (BenTheElder) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:105]: > notably chown doesn't work for Windows, so the current code in this PR is not portable.

none of the control plane components support running on windows to begin with?

only kubelet, containerd or docker, and possibly some of kube-proxy.

Comment (BenTheElder) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:105]: also: `hostPath` volumes do not support altering permissions intentionally, so you must chown them from something else with permission (such as `kubeadm init`), I don't think you can move the pod to rootless without doing this.

Comment (neolit123) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go:105]: commenting in the main thread.

**Comment by neolit123** [comment]
@BenTheElder 

> none of the control plane components support running on windows to begin with?

that may change. the kubelet was arguably the more complicated component to support Windows.

Go's chown is not a portable way for changing ownership.
having a wrapped utility that uses chown on *Unix and a no-op for Windows for the time being seems better to me

@vinayakankugoyal 

https://github.com/kubernetes/kubeadm/issues/1367 needs a KEP
https://github.com/kubernetes/enhancements/blob/master/keps/README.md

TL;DR Kubernetes features first pass a design stage by writing a proposal document called a KEP, instead of being sent as PRs directly. if you want to work on that, you can assign yourself to the k/kubeadm issue and draft a Google doc with your proposal following the KEP template.

some initial feedback from me:
- making only kube-scheduler non-root, will not be accepted.
- the change should ideally be without kubeadm API changes, or if does any they must be scheduled for the next kubeadm API - v1beta3 (timeline unknown).
- we are already using 999/999 as uid/gid in some upgrade pods:
https://github.com/kubernetes/kubernetes/blob/8725c3bf12cfd3697464136201216fa05dc662d2/cmd/kubeadm/app/phases/upgrade/health.go#L123-L124

i'm going to leave this PR open for you if you want to show it as a POC.

**Comment by BenTheElder** [comment]
> that may change. the kubelet was arguably the more complicated component to support Windows

Source? I don't see this changing at all. There are no plans for this that I've seen.

If and when it did change, we'd be running totally different images anyhow, and it would be reasonable to add different codepaths. Worrying about kubeadm init working on windows before windows is even remotely possible while no effort for this is planned seems rather premature.

The images will not be able to run with the UID/GID as well, so there will certainly be new codepaths when windows is added.

Not having a kubeadm API for this means an implicit API for the images (which btw already do not use 999), which is an awkward thing for custom images.

Incrementally moving components is likely much safer than bulk moving all of them.

**Comment by BenTheElder** [comment]
(also this PR is merely a POC, this comes after finishing migrating kube-up apiserver to rootless)

**Comment by neolit123** [comment]
> Source? I don't see this changing at all. There are no plans for this that I've seen.

knowing Windows Server licence buyers, something tells me that this will happen eventually.

> Worrying about kubeadm init working on windows before windows is even remotely possible while no effort for this is planned seems rather premature. (These images / binaries don't even exist yet, and nobody is working on this)

the "chown for the eventual Windows control-planes" argument is poor, i agree.
but a `kubeadmutil.Chown` wrapper is preemptive and can be used for other things too.

> Not having a kubeadm API for this means an implicit API for the images (which btw already do not use 999), which is an awkward thing for custom images.

uid/gid can be hardcoded at first and exposed as an API in v1beta3.

> Incrementally moving components is likely much safer than bulk moving all of them.

this can be part of the KEP. i'd like to see in a proposal what is doable and what not.
and the same proposal then passing security review, even if some things will be obvious.

**Comment by vinayakankugoyal** [comment]
I am going to close this one and we will use a more structured approach to these changes.
